python/ray/data/datasource/dbapi2_datasource.py

```diff
@@ -0,0 +1,122 @@
+import math
+from contextlib import contextmanager
+from functools import lru_cache
+from typing import Any, Callable, Generator, Iterable, List, NewType, Optional
+
+import pyarrow as pa
+
+from ray.data._internal.arrow_block import ArrowRow
+from ray.data.block import Block, BlockAccessor, BlockMetadata
+from ray.data.datasource.datasource import Datasource, Reader, ReadTask, WriteResult
+
+# A Python DB API2-compliant `Connection` object.
+Connection = Any
+
+
+def cursor_to_block(cursor) -> Block:
+    rows = cursor.fetchall()
+    columns = [column_description[0] for column_description in cursor.description]
+    pylist = [{c: v for c, v in zip(columns, row)} for row in rows]
+    return pa.Table.from_pylist(pylist)
```

Comment:
have we checked PyArrow supports all types from those SQL databases? Can you check what error message looks like if encountering a data type that PyArrow not recognize?

Reply:
I think PyArrow should support all of the types.

The DP API2 specification requires string, binary, number, datetime, or row ID (int) types.

To be safe, we could also do something like this:
```
try:
    return pa.Table.from_pylist(...)
except pa.lib.ArrowInvalid as e:
    raise RuntimeError("Your database contains data that Arrow can't recognize.") from e
```